src/mds/MDCache.cc

```diff
@@ -13419,6 +13504,286 @@ void MDCache::clear_dirty_bits_for_stray(CInode* diri) {
+      return;
+    }
+  }
+
+  /* TODO: Consider:
+   *
+   *  rank0 is auth for /foo
+   *  rank1 quiesces /foo with no dirents in cache (and stops)
+   *  rank0 begins quiescing /foo
+   *  rank0 exports a dirfrag of /foo/bar to rank1 (/foo/bar is not authpinned by rank1 nor by rank0 (yet))
+   *  rank1 discovers relevant paths in /foo/bar
+   *  rank1 now has /foo/bar in cache and may issue caps / execute operations
+   *
+   * The solution is probably to have rank1 mark /foo has STATE_QUIESCED and reject export ops from rank0.
+   */
+
+  if (in->is_auth()) {
+    /* Acquire rdlocks on anything which prevents writing.
+     *
+     * Because files are treated specially allowing multiple reader/writers, we
+     * need an xlock here to recall all write caps. This unfortunately means
+     * there can be no readers.
+     *
+     * The xlock on the quiescelock is important to prevent future requests
+     * from blocking on other inode locks while holding path traversal locks.
+     * See dev doc doc/dev/mds_internals/quiesce.rst for more details.
+     */
+
+    MutationImpl::LockOpVec lov;
+    lov.add_rdlock(&in->authlock);
+    lov.add_rdlock(&in->dirfragtreelock);
+    lov.add_rdlock(&in->filelock);
+    lov.add_rdlock(&in->linklock);
+    lov.add_rdlock(&in->nestlock);
+    lov.add_rdlock(&in->policylock);
+    // N.B.: NO xlock/wrlock on quiescelock; we need to allow access to mksnap/lookup
+    // This is an unfortunate inconsistency. It may be possible to circumvent
+    // this issue by having those ops acquire the quiscelock only if necessary.
+    if (is_root) {
+      lov.add_rdlock(&in->quiescelock);
+    } else {
+      lov.add_xlock(&in->quiescelock); /* !! */
+    }
+    lov.add_rdlock(&in->snaplock);
+    lov.add_rdlock(&in->xattrlock);
+    if (!mds->locker->acquire_locks(mdr, lov, nullptr, {in}, false, true)) {
+      return;
+    }
+  } else if (!splitauth) {
+    dout(5) << "auth is split and splitauth is false: " << *in << dendl;
+    qs.add_failed(mdr, -CEPHFS_EPERM);
+    mds->server->respond_to_request(mdr, -CEPHFS_EPERM);
+    return;
+  }
+
+  if (in->get_projected_inode()->get_quiesce_block()) {
+    dout(10) << __func__ << " quiesce is blocked for this inode; dropping locks!" << dendl;
+    mdr->mark_event("quiesce blocked");
+    mds->locker->drop_locks(mdr.get());
+    /* keep authpins! */
+    qs.inc_inodes_blocked();
+    mdr->internal_op_finish->complete(0);
+    mdr->internal_op_finish = nullptr;
+    return;
+  }
+
+  if (in->is_dir()) {
+    for (auto& dir : in->get_dirfrags()) {
+      if (!dir->is_auth() && !splitauth) {
+        dout(5) << "auth is split and splitauth is false: " << *dir << dendl;
+        qs.add_failed(mdr, -CEPHFS_EPERM);
+        mds->server->respond_to_request(mdr, -CEPHFS_EPERM);
+        return;
+      }
+    }
+    uint64_t count = 0;
+    MDSGatherBuilder gather(g_ceph_context, new C_MDS_RetryRequest(this, mdr));
```

Comment:
Just noticed this. Retrying the whole request from the top is too expensive. We should only have here a lambda to complete this request, nothing else.

Reply:
Correction: I had forgotten that it's my POC that performs lock drops. There I really don't want to enter this function from the start once I'm past the lock drop, so it's important to have the completion as a dedicated lambda.

In this version, it's OK to have the mdr retried, the acquire locks should be a no-op, though still an overhead.

Reply:
Implementation details